core/src/main/java/org/apache/iceberg/avro/AvroFileAppender.java

```diff
@@ -77,15 +89,16 @@ public void close() throws IOException {
     if (writer != null) {
       writer.close();
       this.writer = null;
+      isClosed = true;
     }
   }
 
   @SuppressWarnings("unchecked")
   private static <D> DataFileWriter<D> newAvroWriter(
-      Schema schema, PositionOutputStream stream, Function<Schema, DatumWriter<?>> createWriterFunc,
+      Schema schema, PositionOutputStream stream, DatumWriter<?> metricsAwareDatumWriter,
       CodecFactory codec, Map<String, String> metadata) throws IOException {
     DataFileWriter<D> writer = new DataFileWriter<>(
-        (DatumWriter<D>) createWriterFunc.apply(schema));
+        (DatumWriter<D>) metricsAwareDatumWriter);
```

Comment:
Minor: I don't think this needs to be a `MetricsAwareDatumWriter`, right? It isn't in the type signature, so we should name it just `datumWriter`.